src/env/hedged/recover.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+use crate::env::default::LogFile;
+use crate::file_pipe_log::log_file::build_file_reader;
+use crate::file_pipe_log::pipe_builder::FileName;
+use crate::file_pipe_log::reader::LogItemBatchFileReader;
+use crate::file_pipe_log::FileNameExt;
+use crate::internals::parse_reserved_file_name;
+use crate::internals::FileId;
+use crate::internals::LogQueue;
+use crate::Error;
+use std::fs;
+use std::io::Result as IoResult;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use crate::env::log_fd::LogFd;
+use crate::env::DefaultFileSystem;
+use crate::env::{FileSystem, Permission};
+
+use super::util::replace_path;
+
+#[derive(Default)]
+pub(crate) struct Files {
+    pub prefix: PathBuf,
+    pub append_files: Vec<SeqFile>,
+    pub rewrite_files: Vec<SeqFile>,
+    pub reserved_files: Vec<SeqFile>,
+}
+
+pub(crate) enum SeqFile {
+    Path(FileName),
+    Handle((FileName, Arc<LogFd>)),
+}
+
+impl SeqFile {
+    pub fn seq(&self) -> u64 {
+        match self {
+            SeqFile::Path(f) => f.seq,
+            SeqFile::Handle((f, _)) => f.seq,
+        }
+    }
+
+    pub fn path(&self) -> &PathBuf {
+        match self {
+            SeqFile::Path(f) => &f.path,
+            SeqFile::Handle((f, _)) => &f.path,
+        }
+    }
+
+    pub fn remove(&self) -> IoResult<()> {
+        match self {
+            SeqFile::Path(f) => fs::remove_file(&f.path),
+            SeqFile::Handle((f, _)) => fs::remove_file(&f.path),
+        }
+    }
+
+    pub fn copy(&self, file_system: &DefaultFileSystem, to: &PathBuf) -> IoResult<u64> {
+        match self {
+            SeqFile::Path(f) => fs::copy(&f.path, to.as_path()),
+            SeqFile::Handle((_, fd)) => {
+                let mut reader = LogFile::new(fd.clone());
+                let mut writer = LogFile::new(Arc::new(file_system.create(to)?));
+                std::io::copy(&mut reader, &mut writer)
+            }
+        }
+    }
+
+    pub fn into_handle(mut self, file_system: &DefaultFileSystem) -> Self {
+        if let SeqFile::Path(f) = self {
+            let fd = Arc::new(file_system.open(&f.path, Permission::ReadOnly).unwrap());
+            self = SeqFile::Handle((f, fd));
+        }
+        self
+    }
+}
+
+pub(crate) fn catch_up_diff(
```

Comment:
"if the file in to is not in from, delete it"
```suggestion
pub(crate) fn synchronize_files(
```